nimbus/src/enrollment.rs

```diff
@@ -20,28 +23,300 @@ pub enum EnrolledReason {
+                    log::debug!(
+                        "Disqualified experiment enrollment '{}' has been discarded (global opt-out)",
+                        &self.slug
+                    );
+                    // Discard disqualified enrollments on global experiments opt-out.
+                    return Ok(None);
+                }
+                self.clone()
+            }
+            EnrollmentStatus::WasEnrolled { .. } => self.clone(),
+            EnrollmentStatus::Error { .. } => self.clone(), // Cannot recover from errors!
+        }))
+    }
+
+    fn on_experiment_discarded(
+        &self,
+        enrollment_events: &mut Vec<EnrollmentChangeEvent>,
+    ) -> Option<Self> {
+        log::debug!(
+            "Experiment '{}' vanished while we had enrollment status of {:?}",
+            self.slug,
+            self
+        );
+        let (branch, enrollment_id) = match &self.status {
+            EnrollmentStatus::Enrolled {
+                branch,
+                enrollment_id,
+                ..
+            } => (branch, enrollment_id),
+            EnrollmentStatus::Disqualified {
+                branch,
+                enrollment_id,
+                ..
+            } => (branch, enrollment_id),
+            _ => return None, // We were never enrolled, bail.
+        };
+        enrollment_events.push(EnrollmentChangeEvent::new(
+            &self.slug,
+            enrollment_id,
+            branch,
+            None,
+            EnrollmentChangeEventType::Unenrollment,
+        ));
+        Some(Self {
+            slug: self.slug.clone(),
+            status: EnrollmentStatus::WasEnrolled {
+                enrollment_id: enrollment_id.to_owned(),
+                branch: branch.to_owned(),
+                observed_at: {
+                    SystemTime::now()
+                        .duration_since(UNIX_EPOCH)
+                        .expect("Current date before Unix Epoch.")
+                        .as_secs()
+                },
+            },
+        })
+    }
+}
+
 #[derive(Deserialize, Serialize, Debug, Clone, Hash, Eq, PartialEq)]
 pub enum EnrollmentStatus {
     // Enrolled
     Enrolled {
+        enrollment_id: String, // Random ID used for telemetry events correlation.
         reason: EnrolledReason,
         branch: String,
     },
-    // Not enrolled because our evaluator declined to choose us.
-    NotSelected,
-    // Not enrolled because either enrollment is paused or the experiment is over.
-    NotRunning,
-    // Not enrolled because we are not being targeted for this experiment.
-    NotTargeted,
-    // User explicitly opted out.
-    OptedOut,
+    NotEnrolled {
+        reason: NotEnrolledReason,
+    },
+    Disqualified {
+        reason: DisqualifiedReason,
+        enrollment_id: String,
+        branch: String,
+    },
+    WasEnrolled {
+        enrollment_id: String,
+        branch: String,
+        observed_at: u64, // unix timestamp in sec, used to GC old enrollments
```

Comment:
Note: Not implemented yet, that would be a follow-up to this patch.

Reply:
I assume `observed_at` is when we observed we were no longer enrolled? Would changing the name to make that clearer be worthwhile? (OTOH, if it's not implemented, then there's probably a case to be made to just delete it?